### ISSUE TITLE
IronPython: Must remove paths from "Search Paths" and put them back for Intellisense to work

### ISSUE THREAD

**Issue by gnychis** [Microsoft/PTVS#603]
With IronPython support, every time that I close Visual Studio and re-open my project, for Intellisense to work on my Custom .NET DLLs (which are pointed to via my project-based "Search Paths").. I have to remove everything from my project-based "Search Paths" and re-add them for Intellisense to work again.  Otherwise I see `Unable to resolve <something>.  Intellisense may be missing for this module.`  Is PTVS not properly reloading the search paths when loading the project?

Here's a video example where the Intellisense is working perfectly fine and if I simply re-open the project it is not working.  To fix it I have to re-add my search paths.

https://youtu.be/U7gMZTvPZbE

**Comment by zooba** [comment]
>  Is PTVS not properly reloading the search paths when loading the project?

That's certainly what it looks like. We do a few absolute-to-relative path conversions here, and while we have a simple rule (absolute in memory, convert to relative on writing to disk), if we haven't got that quite right here it may be causing this.

**Comment by gnychis** [comment]
Might you be able to point me to a file where the loading is taken care of in PTVS?  I have some time by which I could fork the repository and try and hunt the bug down.

**Comment by zooba** [comment]
If you start in `Python\Product\PythonTools\PythonTools\Project\PythonProjectNode.cs` (don't ask how our hierarchy got so deep...) at the ParseSearchPath functions, you'll see how it's meant to be done. If you then find all uses of the `PythonContants.SearchPathSetting` value, you'll see everywhere else we get search paths directly. Chances are, one of those is not correctly converting to absolute paths.

It's also possible that there's a place that's using `"SearchPath"` to get the paths rather than the constant, so if all of the references to the constant look fine it may be using the string literal instead.

Good luck. :)

**Comment by gnychis** [comment]
Okay great!  This is the exact kind of direction I needed.  Let me see if I can make some headway on it and I'll chime in here if I have any questions. :)

**Comment by gnychis** [comment]
@zooba I just spent the past 5-6 hours trying to debug this and my wheels are starting to spin, so some feedback might help me get back on the right direction.

So far, I am unable to find a single search path in memory that is not the proper full path.  I set breakpoints all around `PythonProjectNodes.cs` and around other parts of the project in general where search path is used, and not once have I found something in memory where it's not an absolute path.

Instead, my suspicion is that the modules are simply not being loaded properly by the remote interpreter when the project is reloaded.  Maybe something happens when freshly loading new search paths, that does not happen when reloading the project?  I don't think a broken search path is causing this, though.

My reasoning behind this is two-fold.  First, in `PythonProjectNode.cs` if I inspect `_analyzer` that `GetAnalyzer()` returns, the _interpreter embedded in the object is missing the _members and _modules.  So it doesn't seem like an intellisense failure, it seems like a failure to load the modules.  Sometimes _assemblyLoadSet in the interpreter is missing my modules also.

Finally, if I add some debugging to `IronPythonInterpreter.cs::ImportModule()` ... I see my custom modules trying to be imported, but they fail to import (null is returned).  I believe it is due to a failure in `Remote.LookupNamespace()`.  I kept trying to step deeper, but at some point `_namespaceTracker.TryGetPackage(name);` is called and its a method I can't look any deeper in to.  It ends up returning `null` though, ultimately leading to my package failing to import.

If I re-add my search paths, IronPythonInterpreter.cs::ImportModule() properly imports my custom module via `Remote.LookupNameSpace()`.

Again, I can't be sure it is a failure in search path... but I haven't been able to find an instance where the search path is not absolute.

I'd greatly appreciate any more guidance!

**Comment by gnychis** [comment]
What also gives me some confidence that it's not search path, is if I modify the .pyproj file and change the search path to be an absolute path it does not fix the issue.

**Comment by zooba** [comment]
That's reasonable.

I just had a quick play, and it looks like if you edit the file after loading it in the editor, the modules are resolved. Does that work for you?

**Comment by gnychis** [comment]
Which file do you edit after loading in the editor? A Python file, or are you talking about the .pyproj file?

**Comment by zooba** [comment]
Just a Python file - mine looks like this:

```python
import clr
clr.AddReference("IEHost")

from Microsoft.IE import Manager
```

Also (for me anyway), none of the code is correctly analysed until you've edited the file. My `import clr` tooltip shows "unknown type" before, and "built-in module clr" afterwards.

**Comment by gnychis** [comment]
Unfortunately that doesn't seem to work for me :\  If I modify the file and save, it still cannot find the modules.

**Comment by gnychis** [comment]
I can try to create a concise example for you.  

But `SetSearchPathEntry()` ends up having the proper paths, and I double checked that `_analysisQueue.Enqueue(...)` such paths was called.  It doesn't seem like I can actually see what happens to items in the analysis queue though.  But I suspect the paths are correct and they are getting passed to the analyzer.  I can't yet track down a difference between the project reloading and the search paths being added.

**Comment by zooba** [comment]
I don't think there is any difference. Rewatching your video, it looks like two problems.

The first is that we rely on a `clr.AddReference` or similar call too actually load the assembly (see `IronPythonInterpreter.SpecializeClrFunctions()` for where this is hooked up). Apparently this is not necessary in some cases, but possibly you've got those calls elsewhere in your project?

Secondly, we aren't analysing the file when you open a project and the file is already open.

Looking at the video, you'll notice that when you remove the search paths, `System.Windows.Forms` is resolved, whereas it starts as unresolved. I think any change to the search paths will trigger reanalysis, as should any change to the file.

In fact, I think the `clr.AddReference` calls are the key. Do you have them in another file?

**Comment by zooba** [comment]
(They should be alright to have in a separate file, but it wouldn't surprise me if we weren't handling that properly. Try opening and changing that file, then changing the original one again.)

**Comment by zooba** [comment]
Okay, I now have two modules in my project.

module1.py:
```python
import clr
clr.AddReference("IEHost")
```

module2.py:
```python
from Microsoft.IE import Manager
```

With both files open and `module2.py` active, I close the solution and then reload it. The module is unresolved. All I have to do is switch to `module1.py` (which loads the file - they are lazily loaded in this context) and then switch back to `module2.py` and everything is resolved correctly.

So for whatever reason, we aren't analysing the entire project correctly when it's loaded. Not sure yet whether it affects more than just IronPython, but this case is certainly reliable.

**Comment by gnychis** [comment]
Very odd, I can't even replicate that behavior of switching between the files or going to another file and saving it.  You're correct in that my `clr.AddReference()` is in another file.  But even if I switch to that file, modify it, and then come back it doesn't resolve the references for me.  I'm trying to create a really concise example for you to check out right now... one sec

**Comment by gnychis** [comment]
Here is a very concise example:  https://github.com/gnychis/PTVSModuleTest

Just build the C# module first, and then open up the IronPython project and you will see the reference is broken.  clr.AddReference() is in the same file and modifying it and saving it doesn't work for me:

https://github.com/gnychis/PTVSModuleTest/blob/master/IPTest/IPTest/IPTest.py

Thanks again for all the help.

**Comment by gnychis** [comment]
Another video showing the issue: https://www.youtube.com/watch?v=jJlOS_U23OE&feature=youtu.be

**Comment by zooba** [comment]
Ah, I see. It looks like when we initially create the `IronPythonInterpreter` instance and hook it up to our `PythonAnalyzer` (via `Initialize()`) we subscribe to the `AnalysisDirectoriesChanged` but we also need to trigger the event then in case it ran before we subscribed. In IronPythonInterpreter.cs I now have this at line 159 and it seems to work okay:

```csharp
            if (_state != null) {
                _state.AnalysisDirectoriesChanged += AnalysisDirectoryChanged;
                AnalysisDirectoryChanged(_state, EventArgs.Empty);
            }
```

Without the direct call to the handler, we weren't adding the search paths into the remote interpreter until they were changed.

This fix won't be in PTVS 2.2, but I'll check it in as soon as we've tagged that release. Thanks for helping track this down.
